kubernetes/cache.go

```diff
@@ -0,0 +1,325 @@
+			return nil, false
+		}
+		nsJobs := make([]batch_v1.Job, len(jobs))
+		for i, job := range jobs {
+			nsJobs[i] = *(job.(*batch_v1.Job))
+		}
+		return nsJobs, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetPods(namespace string) ([]v1.Pod, bool) {
+	indexer := c.controllers["Pod"].GetIndexer()
+	pods, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(pods) > 0 {
+		_, ok := pods[0].(*v1.Pod)
+		if !ok {
+			return nil, false
+		}
+		nsPods := make([]v1.Pod, len(pods))
+		for i, pod := range pods {
+			nsPods[i] = *(pod.(*v1.Pod))
+		}
+		return nsPods, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicationControllers(namespace string) ([]v1.ReplicationController, bool) {
+	indexer := c.controllers["ReplicationController"].GetIndexer()
+	repcons, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repcons) > 0 {
+		_, ok := repcons[0].(*v1.ReplicationController)
+		if !ok {
+			return nil, false
+		}
+		nsRepcons := make([]v1.ReplicationController, len(repcons))
+		for i, repcon := range repcons {
+			nsRepcons[i] = *(repcon.(*v1.ReplicationController))
+		}
+		return nsRepcons, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, bool) {
+	indexer := c.controllers["ReplicaSet"].GetIndexer()
+	repsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repsets) > 0 {
+		_, ok := repsets[0].(*v1beta2.ReplicaSet)
+		if !ok {
+			return nil, false
+		}
+		nsRepsets := make([]v1beta2.ReplicaSet, len(repsets))
+		for i, repset := range repsets {
+			nsRepsets[i] = *(repset.(*v1beta2.ReplicaSet))
+		}
+		return nsRepsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSet(namespace, name string) (*v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, exist, err := indexer.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return nil, false
+	}
+	if exist {
+		fulset, ok := fulsets.(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		return fulset, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(fulsets) > 0 {
+		_, ok := fulsets[0].(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		nsFulsets := make([]v1beta2.StatefulSet, len(fulsets))
+		for i, fulset := range fulsets {
+			nsFulsets[i] = *(fulset.(*v1beta2.StatefulSet))
+		}
+		return nsFulsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetService(namespace, name string) (*v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, exist, err := indexer.GetByKey(namespace + "/" + name)
```

Comment:
Trying to understand how the cache works... is `GetByKey` as fast as `ByIndex` (in other words, keys are also indexed? it's not going to perform full search / O(N) ?)

Reply:
GetByKey() is O(1) as it fetches from the map.
Internally, the cache uses as native keys the same "namespace/name" pattern.
So, before I added a similar index, but then I realized that was not necessary as I can use the native key.